### ISSUE TITLE
Prefered way to extract model hiddens states

### ISSUE THREAD

**Issue by pabloriera** [s3prl/s3prl#394]
Hi, I found at least two ways of loading and calling an upstream model. One is with the hub and the other is with S3PRLUpstream. Testing both ways I found they generate different results in number of hidden states, shapes, and also tensor values

```
wavs = [torch.randn(160000, dtype=torch.float).to(device) for _ in range(2)]
```

```
model = getattr(hub, 'hubert')()
model.eval()
model.to(device)
reps1 = model(wavs)["hidden_states"]
```
len(reps1)==2 and shapes are [2, 499, 768]

```
model = S3PRLUpstream("hubert")
model.eval()
model.to(device)
reps2 = model(wavs, torch.LongTensor([len(w) for w in wavs]))
```
len(reps2[0])==13 and shapes are [2, 501, 768]

I check with np.allclose between the reps1 and the reps2, and I could not match the outputs.

Which of both ways is the prefered way? I tried to dive into the source but could not found the difference. Thanks

**Comment by leo19941227** [comment]
Hi,

The `s3prl.nn.Upstream` is the prefered way.
I believe you should use 

```
reps2, reps2_len = model(wavs, torch.LongTensor([len(w) for w in wavs]))
```

Then you will find the same number of hidden states between reps1 and reps2 (both 13)
Also, by looking at the numerical values they should be the same.
The reason you can not pass the allclose is because they have the different sequence length.
However, usually they are only different for the last 1 ~ 2 frames.
This is because the old s3prl.hub package did not handle the upstream without proper padding. Hence, in s3prl.hub upstreams both in the same `downsample_rate` (frame shift) can have different final sequence length. In `S3PRLUpstream` this behavior is standardized so that you can expect the same sequence length for the models with the same frame shift.

**Comment by pabloriera** [comment]
Perfect, thanks for the clarification. I could match the outputs by selecting the same first number of frames. The hub strategy only exposed the first two hidden state layers, the other layers had a different names.

**Comment by leo19941227** [comment]
Hi,

It seems the issue has been solved. I am closing it.
Feel free to reopen it and raise quesetions.
